libp2p/protocols/relayv2/relayv2.nim

```diff
@@ -0,0 +1,860 @@
+    await handleHopError(conn, RelayV2Status.ResourceLimitExceeded)
+    return
+
+  let src = conn.peerId
+
+  if msg.peer.isNone():
+    await handleHopError(conn, RelayV2Status.MalformedMessage)
+    return
+
+  let dst = msg.peer.get()
+
+  if conn.isCircuitRelay():
+    trace "connection attempt over relay connection", src = conn.peerId
+    await handleHopError(conn, RelayV2Status.PermissionDenied)
+    return
+
+  if dst.peerId notin rv2.rsvp:
+    trace "refusing connection, no reservation", src, dst = dst.peerId
+    await handleHopError(conn, RelayV2Status.NoReservation)
+    return
+
+  rv2.peerCount.inc(src)
+  rv2.peerCount.inc(dst.peerId)
+  defer:
+    rv2.peerCount.inc(src, -1)
+    rv2.peerCount.inc(dst.peerId, -1)
+
+  if rv2.peerCount[src] > rv2.maxCircuitPerPeer or
+     rv2.peerCount[dst.peerId] > rv2.maxCircuitPerPeer:
+    trace "too many connections", src = rv2.peerCount[src],
+                                  dst = rv2.peerCount[dst.peerId],
+                                  max = rv2.maxCircuitPerPeer
+    await handleHopError(conn, RelayV2Status.ResourceLimitExceeded)
+    return
+
+  let connDst = try:
+    await rv2.switch.dial(dst.peerId, RelayV2StopCodec)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error opening relay stream", dst, exc=exc.msg
+    await handleHopError(conn, RelayV2Status.ConnectionFailed)
+    return
+  defer:
+    await connDst.close()
+
+  let stopMsgToSend = StopMessage(msgType: StopMessageType.Connect,
+                          peer: some(RelayV2Peer(peerId: src, addrs: @[])),
+                          limit: some(rv2.limit))
+
+  try:
+    await connDst.writeLp(encodeStopMessage(stopMsgToSend).buffer)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error writing stop handshake", exc=exc.msg
+    await handleHopError(conn, RelayV2Status.ConnectionFailed)
+    return
+
+  let msgRcvFromDstOpt = try:
+    decodeStopMessage(await connDst.readLp(rv2.msgSize))
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error reading stop response", exc=exc.msg
+    await handleHopError(conn, RelayV2Status.ConnectionFailed)
+    return
+
+  if msgRcvFromDstOpt.isNone():
+    trace "error reading stop response", msg = msgRcvFromDstOpt
+    await handleHopError(conn, RelayV2Status.ConnectionFailed)
+    return
+
+  let msgRcvFromDst = msgRcvFromDstOpt.get()
+  if msgRcvFromDst.msgType != StopMessageType.Status:
+    trace "unexpected stop response, not a status message",
+          msgType = msgRcvFromDst.msgType
+    await handleHopError(conn, RelayV2Status.ConnectionFailed)
+    return
+
+  if msgRcvFromDst.status.isNone() or
+     msgRcvFromDst.status.get() != RelayV2Status.Ok:
+    trace "relay stop failure", status = msgRcvFromDst.status
+    await handleHopError(conn, RelayV2Status.ConnectionFailed)
+    return
```

Comment:
Actually, from dial to here could be put in a separate procedure (eg stopPeer) without any try / expect / if isNone
and handle the failure only once